underdocs-version/src/main/kotlin/underdocs/version/reader/VersionPropertiesReader.kt

```diff
@@ -0,0 +1,22 @@
+package underdocs.version.reader
+
+import java.io.BufferedReader
+import java.io.InputStreamReader
+import java.io.StringReader
+import java.util.Properties
+import java.util.stream.Collectors
+
+class VersionPropertiesReader: PropertiesReader {
+    override fun readProperties(): Properties {
+        val properties = Properties()
+        val propertiesFile = "underdocs-version.properties"
+
+        val inputStream = Thread.currentThread().contextClassLoader.getResourceAsStream(propertiesFile)
+        val isr = InputStreamReader(inputStream)
+        val br = BufferedReader(isr)
```

Comment:
When dealing with various resources, such as files, just as you open them, you're also responsible for closing them. Forgetting to do so might result in running out of available file descriptors for example on *nix.

In the JVM world, resources that should be closed implement the `Closeable` or `AutoCloseable` interface. Java also provides the so-called `try-with-resources` construct:

~~~~Java
try (InputStream is = Thread.currentThread().contextClassLoader.getResourceAsStream(propertiesFile)) {
 // is will be automatically closed after this block, even in the case of an exception.
}
~~~~

Although Kotlin does not provide a syntactic construct for this purpose, what it does have, is the `use` extension method: https://www.baeldung.com/kotlin-try-with-resources

In this case, we have multiple closeable things. The following SO answer can provide guidance on this matter: https://stackoverflow.com/questions/884007/correct-way-to-close-nested-streams-and-writers-in-java/21723193#21723193

Reply:
True, forgot about closing them. So I assume something like this would work?

`
val inputStream = Thread.currentThread().contextClassLoader.getResourceAsStream(path)
        inputStream.use {
            val isr = InputStreamReader(inputStream)
            isr.use {
                val br = BufferedReader(isr)
                versionFileContents = br.lines().collect(Collectors.joining("\n"))
            }
        }
`

Reply:
I guess, you also have to close the `BufferedReader`, don't you?

Reply:
True, I need to close that too.